images/kindnetd/cmd/kindnetd/main.go

```diff
@@ -134,43 +181,76 @@ func main() {
-		if nodeIP == "" {
-			klog.Infof("Node %v has no Internal IP, ignoring\n", node.Name)
-			return nil
-		}
-
-		// don't do anything unless there is a PodCIDR
-		podCIDR := node.Spec.PodCIDR
-		if podCIDR == "" {
-			klog.Infof("Node %v has no CIDR, ignoring\n", node.Name)
-			return nil
-		}
-
+		// first get this node's IPs
+		// we don't support more than one IP address per IP family for simplification
+		nodeIPs := internalIPs(node)
+		klog.Infof("Handling node with IPs: %v\n", nodeIPs)
 		// This is our node. We don't need to add routes, but we might need to
-		// update the cni config.
-		if nodeIP == hostIP {
-			klog.Infof("handling current node\n")
+		// update the cni config and "annotate" our external IPs
+		if nodeIPs.Has(hostIP) {
+			klog.Info("handling current node\n")
 			// compute the current cni config inputs
 			if err := cniConfig.Write(
 				ComputeCNIConfigInputs(node),
 			); err != nil {
 				return err
 			}
+			if ipFamily == DualStackFamily {
+				// update external IPs node annotations until #42125 is fixed
+				// TODO: https://github.com/kubernetes/kubernetes/issues/42125
+				// in KIND we know that the outer interface is eth0 to find the external ips
+				// but we can be smarter and find the interface based on the HOST_IP
+				hostExternalIPs := getHostExternalIPs(hostIP)
+				// update the node
+				klog.Infof("Update node with IPs: %v\n", hostExternalIPs)
+				patchString := fmt.Sprintf(`{"metadata": {"annotations": {"kind.x-k8s.io.kindnet/addresses": "%s"}}}`, hostExternalIPs)
+				patchBytes := []byte(patchString)
+				if _, err := clientset.CoreV1().Nodes().Patch(node.Name, types.MergePatchType, patchBytes); err != nil {
+					return fmt.Errorf("failed to patch node: %v", err)
+				}
+			}
 			// we're done handling this node
 			return nil
 		}
 
-		klog.Infof("Handling node with IP: %s\n", nodeIP)
-		klog.Infof("Node %v has CIDR %s \n", node.Name, podCIDR)
-		if err := syncRoute(nodeIP, podCIDR); err != nil {
-			return err
+		// This is another node. Add routes to the POD subnets in the other nodes
+		// don't do anything unless there is a PodCIDR
+		var podCIDRs []string
+		if ipFamily == DualStackFamily {
+			podCIDRs = node.Spec.PodCIDRs
+		} else {
+			podCIDRs = []string{node.Spec.PodCIDR}
+		}
+		if len(podCIDRs) == 0 {
+			fmt.Printf("Node %v has no CIDR, ignoring\n", node.Name)
+			return nil
+		}
+		klog.Infof("Node %v has CIDR %s \n", node.Name, podCIDRs)
+		podCIDRsv4, podCIDRsv6 := splitCIDRs(podCIDRs)
+
+		// obtain the PodCIDR gateway
+		var nodeIPv4, nodeIPv6 string
+		for _, ip := range nodeIPs.List() {
+			if utilsnet.IsIPv6String(ip) {
+				nodeIPv6 = ip
+			} else {
+				nodeIPv4 = ip
+			}
 		}
 
+		if nodeIPv4 != "" && len(podCIDRsv4) > 0 {
+			if err := syncRoute(nodeIPv4, podCIDRsv4); err != nil {
+				return err
+			}
+		}
+		if nodeIPv6 != "" && len(podCIDRsv6) > 0 {
+			if err := syncRoute(nodeIPv6, podCIDRsv6); err != nil {
+				return err
+			}
+		}
```

Comment:
we are not handling most errors in kindnetd, to try to keep it as simple as possible, it just simple keep retrying
We can take this approach because we have control of the network scenario with docker/podman, so if any node does not have the corresponding IP family address is mostly a docker/podman bug and more things will break at the deployment before kindnet.